Updated the api to handle request

breeshidy opened the pull request:
Problem: Update API to handle request to `/api/cohorts?sort=date&order=descending`

Solution: Added a Get Route to handle requests by passing the params

Issue: 
- if you do not pass the "order" query it will default to "ASC" order
- if you do not pass the "sort" query it will default to "startDate"
- So the default parameters are "sort=startDate&orde=ASC" 
- if you add invalid query parameters,  it will return an error message saying "Sorry invalid query parameters" 


Jira: https://dfacademy.atlassian.net/browse/AT-9

dearshrewdwit commented on cypress/integration/cohorts.spec.js:
```diff
@@ -1,16 +1,29 @@
+    });
+  });
+
+    it("returns cohorts to /api/cohorts?sort=date&order=descending", function () {
```
How is this one test ensuring the sort and order?

I suggest creating a few cohorts, and having a few tests to sort via name, date and order by ASC, DESC.

breeshidy replied:
This has now been done

dearshrewdwit commented on cypress/fixtures/cohorts.json:
```diff
@@ -0,0 +1,25 @@
+{
```
What's the purpose of this file?

breeshidy replied:
This was us playing our with the test -- we will be sure to add it to our gitignore or probably delete it :)

dearshrewdwit replied:
You marked this conversation as resolved, but the file is still here. What's the reasoning to keep it?

dearshrewdwit commented on app.js:
```diff
@@ -1,18 +1,18 @@
-const app = express()
-const port = 5000
-const cors = require('cors')
+const express = require("express");
```
This is a nodejs style change - if so, let's put this on a separate PR, use eslint, and enforce it!

Otherwise this will cause lots of changes to people's code as they go from single to double quotes and back again.

So for this PR, undo these style changes.

breeshidy replied:
sorry it's the style of my vs code that automatically changed it but I have undone the changes now

dearshrewdwit commented on package.json:
```diff
@@ -21,5 +20,8 @@
     "sequelize": "^6.6.5",
     "sequelize-cli": "^6.2.0"
+  },
+  "devDependencies": {
```
Nice. Let's put this on a separate PR though.

breeshidy replied:
i am not sure what you mean -- should I create a pull request for updating cypress dependency?

dearshrewdwit replied:
Yes, correct. It's not part of this bit of work.

dearshrewdwit commented on routes/cohorts.js:
```diff
@@ -1,12 +1,30 @@
 
-router.get('/', async function (req, res) {
-  const cohorts = await Cohort.findAll({ include: { all: true }})
+router.get("/", async function (req, res) {
```
Why are there two routes matching `GET` and `"/"` ?

breeshidy replied:
Sorry forgot to delete it, we were just testing the difference

dearshrewdwit commented on migrations/20210817094012-create-added-date-field-to-cohorts.js:
```diff
@@ -0,0 +1,14 @@
+  up: async (queryInterface, Sequelize) => {
+     await queryInterface.addColumn(
+     'Cohorts',
+     'date',
```
Can we name this column `startDate` to avoid ambiguity

breeshidy replied:
yes, sure can

dearshrewdwit reviewed (changes requested):
See below!

dearshrewdwit commented:
@breeshidy if this needs a review, please share in code-review in discord 👍

breeshidy commented:
> @breeshidy if this needs a review, please share in code-review in discord 👍

Done

dearshrewdwit commented on routes/cohorts.js:
```diff
@@ -1,12 +1,48 @@
+  console.log("sort " + sort + " order " + order)
+  console.log("req.params " + req.query.order )
+
+  if(sort === undefined && order === undefined){
```
This if statement can be DRYed up.

How can you set some default values for `sort` and `order` so you can avoid the if statement?

dearshrewdwit commented on test/create-cohort.js:
```diff
@@ -3,7 +3,7 @@ const { Cohort } = require('../models')
   console.log('creating cohort')
   await Cohort.create({
-    name: 'test cohort'
+    name: 'test cohort',
```
What's the reasoning to add a `,` here?

breeshidy replied:
wanted to add the start date field

dearshrewdwit reviewed (changes requested):
Good. Some more work needed.

See comments below.

breeshidy commented on routes/cohorts.js:
```diff
@@ -1,12 +1,48 @@
+router.get("/", async function (req, res) {
+  const sort = req.query.sort;
+  const order = req.query.order;
+  console.log("sort " + sort + " order " + order)
```
Done

breeshidy replied:
Done

breeshidy commented on package.json:
```diff
@@ -22,4 +22,4 @@
     "sequelize-cli": "^6.2.0"
   }
-}
+}
```
i guess i mistakenly took the } out

breeshidy commented on cypress/integration/cohorts.spec.js:
```diff
@@ -1,16 +1,85 @@
-    cy.task('taskTruncateTables')
-      .then(() => {
-        cy.task('taskCreateCohort')
+describe("Cohorts", function () {
```
updated test to add 3 different cohorts

dearshrewdwit commented on routes/cohorts.js:
```diff
@@ -1,12 +1,19 @@
-  const cohorts = await Cohort.findAll({ include: { all: true }})
+  const sort = req.query.sort || "startDate"
+  const order = req.query.order || "ASC"
+  console.log('sort ' + sort + ' order ' + order)
```
Let's remove debugging statements

breeshidy replied:
yes,  sorry missed that

breeshidy commented:
> Good. Some more work needed.
> 
> See comments below.

breeshidy commented:
> Good. Some more work needed.
> 
> See comments below.

Hello, @dearshrewdwit, the changes are made now. Do I submit on code review again?

dearshrewdwit commented on test/create-cohort.js:
```diff
@@ -3,7 +3,16 @@ const { Cohort } = require('../models')
   await Cohort.create({
-    name: 'test cohort'
+    name: 'A test',
+    startDate: '12/01/2021'
```
What format is this? 'DD/MM/YYYY' or 'MM/DD/YYYY'? 

How do you know?

breeshidy replied:
Ehh according to Postgres -- the format is YYYY-MM-DD format, is this the right format? I will change the test to match it

dearshrewdwit replied:
Yes, but how does `'12/01/2021'` get parsed into Postgres' format YYYY-MM-DD?

dearshrewdwit replied:
Get visibility from data in your table or from sequelize after you've created the these records.

breeshidy replied:
Okay will adjust that now

dearshrewdwit reviewed (changes requested):
Nice. See comments below

dearshrewdwit commented:
> > Good. Some more work needed.
> > See comments below.
> 
> Hello, @dearshrewdwit, the changes are made now. Do I submit on code review again?

@breeshidy yes, whenever you are looking for a review, share back in code-review 👍

dearshrewdwit reviewed (approved):
Ok. Good enough. @breeshidy 

1. Add an issue to your board to clarify what happens if you pass request parameters that are not valid. eg: `?sort=XYZ&order=ABC` (it might be nothing, and simply reverts to the default behaviour. This should be documented.

2. In future please force squash your commits -> https://github.com/digital-futures-academy/atlas-api/pull/9/commits so that it reads easily as the history of change for this piece of work - currently it's hard to understand what happened!

breeshidy commented:
> Ok. Good enough. @breeshidy
> 
> 1. Add an issue to your board to clarify what happens if you pass request parameters that are not valid. eg: `?sort=XYZ&order=ABC` (it might be nothing, and simply reverts to the default behaviour. This should be documented.
> 2. In future please force squash your commits -> https://github.com/digital-futures-academy/atlas-api/pull/9/commits so that it reads easily as the history of change for this piece of work - currently it's hard to understand what happened!



> 2\. please force squash your commits

Hello

> Ok. Good enough. @breeshidy
> 
> 1. Add an issue to your board to clarify what happens if you pass request parameters that are not valid. eg: `?sort=XYZ&order=ABC` (it might be nothing, and simply reverts to the default behaviour. This should be documented.
> 2. In future please force squash your commits -> https://github.com/digital-futures-academy/atlas-api/pull/9/commits so that it reads easily as the history of change for this piece of work - currently it's hard to understand what happened!

Hello Ed, I have now added the issue section now. 

> Ok. Good enough. @breeshidy
> 
> 1. Add an issue to your board to clarify what happens if you pass request parameters that are not valid. eg: `?sort=XYZ&order=ABC` (it might be nothing, and simply reverts to the default behaviour. This should be documented.
> 2. In future please force squash your commits -> https://github.com/digital-futures-academy/atlas-api/pull/9/commits so that it reads easily as the history of change for this piece of work - currently it's hard to understand what happened!

1-> I have added the issue section of the code and updated it to catch the error and test for invalid inputs

2-> In terms of the squash command. I have tried to google it but I do not seem to understand it. please could you provide steps on how to use it?

dearshrewdwit commented on routes/cohorts.js:
```diff
@@ -9,9 +9,13 @@ router.get('/', async function (req, res) {
+  }).then(cohorts =>{
+    res.json({ cohorts: cohorts })
+  }).catch(errors => {
+    res.json({ errors: "Sorry invalid query parameters" })
```
Where's the test for this error message?

dearshrewdwit replied:
Remember: on the front end you'll need to have a component that can handle a response object like this

breeshidy replied:
we have added the test and it will return an array of errors

breeshidy replied:
@dearshrewdwit , I think I messed up something with my commit, please can you have a look and see?

dearshrewdwit reviewed (changes requested):
Almost there.

Remember - design behaviour, **codify it** in a **test**, and then write source code.

dearshrewdwit commented:
@breeshidy there were two approvals from me and @Reginmaru .... 
> breeshidy dismissed stale reviews from Reginmaru and dearshrewdwit via 5b9bdad 16 minutes ago
